torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1129,6 +1129,102 @@ def _raise_on_use_orig_params_optim_checkpoint(
 
+    @staticmethod
+    def _optim_state_dict_impl(
+        model: torch.nn.Module,
+        optim: torch.optim.Optimizer,
+        optim_state_dict: Dict[str, Any],
+        optim_input: Optional[
+            Union[
+                List[Dict[str, Any]],
+                Iterable[torch.nn.Parameter],
+            ]
+        ] = None,
+        rank0_only: bool = True,
+        group: Optional[dist.ProcessGroup] = None,
+        full_state_dict: bool = True,
+    ) -> Dict[str, Any]:
+        """
+        The internal API that is used by all the optim_state_dict implementations.
+        """
+        if full_state_dict:
+            FullyShardedDataParallel._raise_on_use_orig_params_optim_checkpoint(
+                model, True, rank0_only
+            )
+            FullyShardedDataParallel._warn_optim_input(optim_input)
+            using_optim_input = FullyShardedDataParallel._is_using_optim_input(
+                optim_input,
+                optim,
+            )
+        else:
+            using_optim_input = False
+            assert optim_input is None and not rank0_only
+            FullyShardedDataParallel._raise_on_use_orig_params_optim_checkpoint(
+                model, False, False
+            )
+
+        use_orig_params = False
+        for module in FullyShardedDataParallel.fsdp_modules(model):
+            use_orig_params = module._use_orig_params
+            break
+
+        return _optim_state_dict(
+            model=model,
+            optim=optim,
+            optim_state_dict=optim_state_dict,
+            optim_input=optim_input,
+            rank0_only=rank0_only,
+            shard_state=not full_state_dict,
+            group=group,
+            using_optim_input=using_optim_input,
+            use_orig_params=use_orig_params,
+        )
+
+    @staticmethod
+    def _optim_state_dict_to_load_impl(
+        optim_state_dict: Dict[str, Any],
+        model: torch.nn.Module,
+        optim_input: Optional[
+            Union[
+                List[Dict[str, Any]],
+                Iterable[torch.nn.Parameter],
+            ]
+        ] = None,
+        optim: Optional[torch.optim.Optimizer] = None,
+        full_state_dict: bool = True,
+        is_named_optimizer: bool = False,
+    ) -> Dict[str, Any]:
+        FullyShardedDataParallel._raise_on_use_orig_params_optim_checkpoint(
+            model, full_state_dict, False
+        )
+        FullyShardedDataParallel._warn_optim_input(optim_input)
+        using_optim_input = FullyShardedDataParallel._is_using_optim_input(
+            optim_input,
+            optim,
+        )
+
+        use_orig_params: bool = False
+        for module in FullyShardedDataParallel.fsdp_modules(model):
+            use_orig_params = module._use_orig_params
+            break
+
+        sharded_osd = _flatten_optim_state_dict(
+            optim_state_dict,
+            model,
+            True,
+            use_orig_params,
+        )
+        if is_named_optimizer:
```

Comment:
might be useful to have a small comment here saying that NamedOptim expects the keys to be FQNs instead of regular optimizers.

Reply:
Will add this in the next PR.